extra/modules/fiftyone-devicedetection/src/main/java/org/prebid/server/hooks/modules/fiftyone/devicedetection/v1/core/DeviceEnricher.java

```diff
@@ -0,0 +1,262 @@
+            return patchDevice(properDevice, deviceData);
+        }
+    }
+
+    private Map<String, String> pickRelevantFrom(CollectedEvidence collectedEvidence) {
+        final Map<String, String> evidence = new HashMap<>();
+
+        final String ua = collectedEvidence.deviceUA();
+        if (StringUtils.isNotBlank(ua)) {
+            evidence.put("header.user-agent", ua);
+        }
+        final Map<String, String> secureHeaders = collectedEvidence.secureHeaders();
+        if (MapUtils.isNotEmpty(secureHeaders)) {
+            evidence.putAll(secureHeaders);
+        }
+        if (!evidence.isEmpty()) {
+            return evidence;
+        }
+
+        Stream.ofNullable(collectedEvidence.rawHeaders())
+                .flatMap(Collection::stream)
+                .forEach(rawHeader -> evidence.put("header." + rawHeader.getKey(), rawHeader.getValue()));
+
+        return evidence;
+    }
+
+    private EnrichmentResult patchDevice(Device device, DeviceData deviceData) {
+        final List<String> updatedFields = new ArrayList<>();
+        final Device.DeviceBuilder deviceBuilder = device.toBuilder();
+
+        final Integer currentDeviceType = device.getDevicetype();
+        if (!isPositive(currentDeviceType)) {
+            final String rawDeviceType = getSafe(deviceData, DeviceData::getDeviceType);
+            if (rawDeviceType != null) {
+                final Integer properDeviceType = convertDeviceType(rawDeviceType);
+                if (isPositive(properDeviceType)) {
+                    deviceBuilder.devicetype(properDeviceType);
+                    updatedFields.add("devicetype");
+                }
+            }
+        }
+
+        final String currentMake = device.getMake();
+        if (StringUtils.isBlank(currentMake)) {
+            final String make = getSafe(deviceData, DeviceData::getHardwareVendor);
+            if (StringUtils.isNotBlank(make)) {
+                deviceBuilder.make(make);
+                updatedFields.add("make");
+            }
+        }
+
+        if (StringUtils.isBlank(device.getModel())) {
+            final String model = getSafe(deviceData, DeviceData::getHardwareModel);
+            if (StringUtils.isNotBlank(model)) {
+                deviceBuilder.model(model);
+                updatedFields.add("model");
+            } else {
+                final List<String> names = getSafe(deviceData, DeviceData::getHardwareName);
+                if (CollectionUtils.isNotEmpty(names)) {
+                    deviceBuilder.model(String.join(",", names));
+                    updatedFields.add("model");
+                }
+            }
+        }
+
+        if (StringUtils.isBlank(device.getOs())) {
+            final String os = getSafe(deviceData, DeviceData::getPlatformName);
+            if (StringUtils.isNotBlank(os)) {
+                deviceBuilder.os(os);
+                updatedFields.add("os");
+            }
+        }
+
```

Comment:
What about to extract each of this patch part to separate method like this?
```java
private UpdateResult<Integer> resolveDeviceType(Device device, DeviceData deviceData) {
        final Integer currentDeviceType = device.getDevicetype();
        if (isPositive(currentDeviceType)) {
            return UpdateResult.unaltered(currentDeviceType);
        }

        final String rawDeviceType = getSafe(deviceData, DeviceData::getDeviceType);
        if (rawDeviceType == null) {
            return UpdateResult.unaltered(currentDeviceType);
        }

        final Integer properDeviceType = convertDeviceType(rawDeviceType);
        return isPositive(properDeviceType)
                ? UpdateResult.updated(properDeviceType)
                : UpdateResult.unaltered(currentDeviceType);
    }
```

```java
    private EnrichmentResult patchDevice(Device device, DeviceData deviceData) {
        final List<String> updatedFields = new ArrayList<>();
        final Device.DeviceBuilder deviceBuilder = device.toBuilder();

        final UpdateResult<Integer> resolvedDeviceType = resolveDeviceType(device, deviceData);
        if (resolvedDeviceType.isUpdated()) {
            deviceBuilder.devicetype(resolvedDeviceType.getValue());
            updatedFields.add("devicetype");
        }

       // and so on
```

Reply:
:heavy_check_mark: Extracted.